File: oop-exercises/inheritance/nestjs/src/entity/SavingsAccount.ts
export class SavingsAccount {
  private balance: number;
  private interestRate: number;

  constructor(balance: number, interestRate: number) {
    this.balance = balance;
    this.interestRate = interestRate;
  }

  public deposit(amount: number): void {
    this.balance += amount;
  }

  public withdraw(amount: number): void {
    if (this.balance < amount) {
      throw new Error("Insufficient balance to perform the transaction.");
    }
    this.balance -= amount;
  }

  public calculateInterest(): number {
    return this.balance * this.interestRate;
  }

  public getBalance(): number {
    return this.balance;
  }

  public getInterestRate(): number {
    return this.interestRate;
  }
}

File: oop-inheritance/spring-boot/src/main/java/com/juandaabril/entity/SavingsAccount.java
package com.juandaabril.entity;

public class SavingsAccount {
  private double balance;
  private double interestRate;

  public SavingsAccount(double balance, double interestRate) {
    this.balance = balance;
    this.interestRate = interestRate;
  }

  public void deposit(double amount) {
    balance += amount;
  }

  public void withdraw(double amount) {
    if (balance < amount) {
      throw new RuntimeException("Insufficient balance to perform the transaction.");
    }
    balance -= amount;
  }

  public double calculateInterest() {
    return balance * interestRate;
  }

  public double getBalance() {
    return balance;
  }

  public double getInterestRate() {
    return interestRate;
  }
}
